SeqManager: Properly remove dropped entries

jmillan opened the pull request:
Fixes #1037.

Keep dropped entries clean by removing those higher than this->maxInput.

This cleans both the `Insert()` and `ClearDropped()` methods. We remove those dropped entries which are from a previous sequence iteration.


STR with mediasoup-demo with single host:

Tab 1: roomId=loss&forceVP9=true&consume=false
Tab 2: roomId=loss&forceVP9=true&produce=false

In tab 2, select the preferred layers (0,0). Many packets are discarded by the codec

It takes few minutes until the seq number wraps ups. Wait for 1 or two wraps up. Vide in Tab 2 should be OK.

jmillan committed:
SeqManager: Properly remove dropped entries

Fixes #1037.

Keep dropped entries clean by removing those higher than this->maxInput.

jmillan commented:
@vpalmisano I've tested this with the STR that I provided and which is failing in V3. Can you give it a try?

jmillan committed:
cosmetic

ibc commented:
I must confess that I don't understand whether the PR description describes an issue or the working solution.

vpalmisano commented:
> Can you give it a try?

After 1h test I got no big loss events, so probably we could say that this was fixed :)

jmillan commented:
Test failing.. I was filtering by [SeqManager] in local..., checking..

jmillan committed:
Merge remote-tracking branch 'origin/v3' into SeqManagerFix

jmillan committed:
Set dropped input to 'maxInput' if needed

A dropped input is part of the same sequence, hence set it to 'maxInput'
if the dropped value is higher than the current 'maxInput'.

Without this consideration we would be wrongly erasing dropped inputs
that are legitimally just the maximum input of the sequence stream.

jmillan commented:
There are some changes in the tests generated out of the Fuzzer output. I've reviewed them and they make sense.

jmillan committed:
Update CHANGELOG

ibc commented on worker/src/RTC/SeqManager.cpp:
```diff
@@ -70,6 +71,14 @@ namespace RTC
 		// There are dropped inputs. Synchronize.
 		if (!this->dropped.empty())
 		{
+			// Set maxInput here if needed before calling ClearDropped.
```
```suggestion
			// Set maxInput here if needed before calling ClearDropped().
```

jmillan committed:
Fix typo

jmillan committed:
Enhance comments

jmillan committed:
Input(): Performance, do not loop over every dropped inputs

Which has a linear complexity. Use lower_bound which has logarithmic
complexity.

jmillan commented:
@ibc, no hurries. Reassigning for review since I've done some performance improvements. IMHO `Input()` method is now even more clear and understandable:

1: No need to consider dropped inputs if `this->dropped` is empty.
2: If 'input'  is higher than `this->maxInput` then reassign `this->maxInput` and clean dropped inputs accordingly.
3: No need to consider dropped inputs if `this->dropped` is empty after cleanup.
4: Nothing to do if 'input' is within `this->dropped`.
5: Calculate new `base` if there are dropped inputs lower than 'input'.

ibc reviewed (approved):
LGTM. Is it fuzzer safe?

jmillan commented:
> LGTM. Is it fuzzer safe?

I didn't try honestly. let me try it tomorrow morning.

jmillan commented:
>> LGTM. Is it fuzzer safe?

> I didn't try honestly. let me try it tomorrow morning.

We are safe.
